engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/PatchedModuleValues.java

```diff
@@ -0,0 +1,133 @@
+package org.enso.interpreter.runtime.scope;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.instrumentation.EventBinding;
+import com.oracle.truffle.api.instrumentation.EventContext;
+import com.oracle.truffle.api.instrumentation.ExecutionEventListener;
+import com.oracle.truffle.api.instrumentation.Instrumenter;
+import com.oracle.truffle.api.instrumentation.SourceFilter;
+import com.oracle.truffle.api.instrumentation.SourceSectionFilter;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.source.SourceSection;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.TreeMap;
+import org.enso.interpreter.runtime.Module;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.tag.Patchable;
+import org.enso.polyglot.LanguageInfo;
+import org.enso.text.editing.model;
+
+final class PatchedModuleValues implements ExecutionEventListener {
+  private final TreeMap<Integer,int[]> deltas = new TreeMap<>();
+  private final Map<Node, Object> values = new HashMap<>();
+  private final EventBinding<PatchedModuleValues> binding;
+
+  PatchedModuleValues(Instrumenter instr, Module module) {
+    SourceFilter sourceAssociatedWithMyModule = SourceFilter.newBuilder().languageIs(LanguageInfo.ID).sourceIs(t -> {
+      return module.isModuleSource(t);
+    }).build();
+    SourceSectionFilter filter = SourceSectionFilter.newBuilder().sourceFilter(sourceAssociatedWithMyModule).tagIs(Patchable.Tag.class).build();
+    this.binding = instr.attachExecutionEventListener(filter, this);
+  }
+
+  void dispose() {
+    binding.dispose();
+  }
+
+  void registerValues(Map<Node, Object> collect) {
+    values.putAll(collect);
+  }
+
+  static void updateFunctionsMap(model.TextEdit edit, Collection<? extends Map<?, Function>> values, Map<Node, Object> nodeValues) {
+    for (Map<?, Function> map : values) {
+      for (Function f : map.values()) {
+        updateNode(edit, f.getCallTarget().getRootNode(), nodeValues);
+      }
+    }
+  }
+
+  private static void updateNode(model.TextEdit edit, Node root, Map<Node, Object> nodeValues) {
```

Comment:
This logic seems really awkward here. Why would an instrument be responsible for computing node positions?

Reply:
The `ExecutionEventListener` wants to modify return values for certain `Patchable` nodes. It uses textual positions to locate the right nodes and put them into `values`map. It has to be ready for the underlying text changes - e.g. it also maintains the _deltas_, so subsequent updates find the right nodes again. All this functionality is only needed because we are _patching module values_ and as such it is handled in the `PatchedModuleValues` class. True, making the class also implement `ExecutionEventListener` is against encapsulation - I could turn the listener into a separate class - but the effect on behavior would be none.